pylint/checkers/variables.py

```diff
@@ -1078,31 +1071,7 @@ def open(self) -> None:
             "undefined-loop-variable"
         )
 
-    @utils.only_required_for_messages("redefined-outer-name")
-    def visit_for(self, node: nodes.For) -> None:
-        assigned_to = [a.name for a in node.target.nodes_of_class(nodes.AssignName)]
-
-        # Only check variables that are used
-        dummy_rgx = self.linter.config.dummy_variables_rgx
-        assigned_to = [var for var in assigned_to if not dummy_rgx.match(var)]
-
-        for variable in assigned_to:
-            for outer_for, outer_variables in self._loop_variables:
-                if variable in outer_variables and not in_for_else_branch(
-                    outer_for, node
-                ):
-                    self.add_message(
-                        "redefined-outer-name",
-                        args=(variable, outer_for.fromlineno),
-                        node=node,
-                    )
-                    break
-
-        self._loop_variables.append((node, assigned_to))
-
-    @utils.only_required_for_messages("redefined-outer-name")
     def leave_for(self, node: nodes.For) -> None:
-        self._loop_variables.pop()
```

Comment:
I don't see the initialization of ``_loop_variables`` being removed, was it used for something else too ?

Reply:
The `.pop()` was moved to the optional checker, which still makes use of the _loop_variables (initialized somewhere else)

Reply:
Essentially, `redefined-outer-name` was doing two things, including one special case for nested loops. I moved the logic for that special-nested-loops case into the new, more robust optional checker.

Reply:
So ``_loop_variables `` is used in both checkers now? Is it populated the same way in both? Won't the lack of pop cause problem in the old checker ?

Reply:
No, it's no longer used in the old variables checker. Sorry this MR took a lot of twists and turns. It's only needed for `redefined-loop-name` so it sits in the new optional checker for it. That new optional checker now has `visit_assignname` and `visit_for` implemented, both of which need access to `_loop_variables`. One of those two methods is new functionality (#5072), the second one is moved from the variables checker where it was mixed in with `redefined-outer-name` and confusing the reporter of #5608

Reply:
Oh! you mean the `= []`. Duh. Yeah I can get rid of it.

Reply:
Sorry I could have been clearer here 😄